lib/stream/GenericWorker.js

```diff
@@ -0,0 +1,227 @@
+ *
+ * A chunk is an object with 2 attributes : `meta` and `data`. The former is an
+ * object containing anything (`percent` for example), see each worker for more
+ * details. The latter is the real data (String, Uint8Array, etc).
+ *
+ * @constructor
+ * @param {String} name the name of the stream (mainly used for debugging purposes)
+ */
+function GenericWorker(name) {
+    // the name of the worker
+    this.name = name || "default";
+    // an object containing metadata about the workers chain
+    this.streamInfo = {};
+    // an error which happened when the worker was paused
+    this.generatedError = null;
+    // an object containing metadata to be merged by this worker into the general metadata
+    this.extraStreamInfo = {};
+    // true if the stream is paused (and should not do anything), false otherwise
+    this.isPaused = true;
+    // true if the stream is finished (and should not do anything), false otherwise
+    this.isFinished = false;
+    // the event listeners
+    this._listeners = {
+        'data':[],
+        'end':[],
+        'error':[]
+    };
+    // the previous worker, if any
+    this.previous = null;
+}
+
+GenericWorker.prototype = {
+    /**
+     * Push a chunk to the next workers.
+     * @param {Object} chunk the chunk to push
+     */
+    push : function (chunk) {
+        this.emit("data", chunk);
+    },
+    /**
+     * End the stream.
+     * @return {GenericWorker} the current object for chainability
+     */
+    end : function () {
+        if (this.isFinished) {
+            return this;
+        }
+
+        this.flush();
+        try {
+            this.emit("end");
+            this.cleanUp();
+            this.isFinished = true;
+        } catch (e) {
+            this.emit("error", e);
+        }
+        return this;
+    },
+    /**
+     * End the stream with an error.
+     * @param {Error} e the error which caused the premature end.
+     * @return {GenericWorker} the current object for chainability
+     */
+    error : function (e) {
+        if (this.isFinished) {
+            return this;
+        }
+
+        if(this.isPaused) {
+            this.generatedError = e;
+        } else {
+            this.emit("error", e);
+            this.cleanUp();
+            this.isFinished = true;
+        }
+        return this;
+    },
+    /**
+     * Add a callback on an event.
+     * @param {String} name the name of the event (data, end, error)
+     * @param {Function} listener the function to call when the event is triggered
+     * @return {GenericWorker} the current object for chainability
+     */
+    on : function (name, listener) {
+        this._listeners[name].push(listener);
+        return this;
+    },
+    /**
+     * Abort the current worker and the upstream workers because of a write
+     * failure (something went wrong downstream and the whole chain explodes).
+     */
+    writeError : function () {
+        this.isFinished = true;
+        this.cleanUp();
```

Comment:
This method is called when something goes wrong in the `ZipFileWorker`. I don't remember why I don't just call `end()` (or why I don't propagate the `end` event upward), I'll investigate.